### DISCUSSION TITLE
HPC cluster host for SSH certificate/SSO authentication, possible to setup as a user and not admin?

### DISCUSSION THREAD

**Discussion by mshamash** [smallstep/certificates#443] [Q&A] [answered]
Hi,

I’m a new user to Smallstep and all the associated cli tools and am looking into the best way to implement step SSH certificates with SSO amongst the many hosts I connect to regularly.

One of these hosts is particularly unique, however, as it is a high performance computing (HPC) cluster where I am only one of hundreds (or thousands?!) of users who connects regularly via SSH. Ultimately, this means that I don’t have permission to install custom applications or scripts, with rare exceptions, and definitely cannot modify global SSH configs in /etc/ and other “core Linux” directories. 
I was looking at the guide Smallstep published on manual host setup (https://smallstep.com/docs/ssh/hosts-step-by-step) and it looks like I ultimately need be sysadmin to enable step integration on a host, is this correct? I was wondering if there were any alternative solutions in this situation, as I’d love to use my newly-built “Tiny CA” in my homelab for SSH certificates everywhere, including the shared HPC cluster (thanks @tashian for the awesome guide you wrote up!).
I feel I may have to just end up going the boring “old-fashioned” route and use an OpenSSL certificate I generate locally for this host in particular...though I’d love to be proven wrong!

Looking forward to hearing your thoughts on this, and best wishes for the New Year to the Smallstep team!

Cheers,

Michael

**Comment by maraino** [comment]
Hi @mshamash, the docs you link are part of our SaaS offering https://smallstep.com/sso-ssh/, it allows you to link your identity provider (IdP), and give access to the users in your IdP to your servers with some policy.

To be able to use certificates you need to change /etc/ssh/sshd_config, you need three parameters, for example:
```
TrustedUserCAKeys /etc/ssh/ca.pub
HostCertificate /etc/ssh/ssh_host_ed25519_key-cert.pub
HostKey /etc/ssh/ssh_host_ed25519_key
```
**TrustedUserCAKeys** defines the file containing the public part of the key used to sign the user certificates.

**HostCertificate** is the SSH certificate for that host, users will a `@cert-authority` entry in their `known_hosts` with the public part of the key used to sign host certificates.

**HostKey** is the private key used by SSH, there are some defaults, and you can omit it if you use one of the defaults.

Our open-source CA allows you to create the host, and user certificates, the easiest way to do it is to bootstrap your CA using 

```sh
step ca init --ssh
```

That will command will create root and intermediate certificates for TLS as well as user and host key for signing SSH certificates. It will also create some templates to automatically configure users using `step ssh config` and hosts using `step ssh config --host`. You can create host certificates for an existing key using:

```sh 
# step ssh certificate -f --sign --host <hostname> <path-to-ssh-public-key>
step ssh certificate -f --sign --host tynyca.local /etc/ssh/ssh_host_ed25519_key.pub
```

Once you configure your hosts you will need an automatic way to refresh the host certificates, by default they expire after 30 days, you can do it with a cron or a systemd timer. [Here](https://smallstep.com/docs/step-ca/certificate-authority-server-production#-systemd-based-renewal) are some docs, but they are for X.509 (TLS) certificates instead of SSH ones. At the end of [this script](https://gist.github.com/tashian/fde43668cbf6e3227fb13ef51db650b8), there's an example to configure crond for this.

The default certificate for users is 16 hours, the best way to refresh them is configuring an IdP and using [OAuth/OIDC](https://smallstep.com/docs/step-ca/configuration#oauthoidc-single-sign-on) although you can also use the default JWK provisioner created in the bootstrap. The configuration files created using `step ssh config` can detect if you need a certificate and redirect you to the authentication flow.

The other problem to solve is the user accounts in the hosts, if you already have one, then there is no problem, you should be able to login with your certificate, if not, then you need a way to provision it, that's where our SaaS comes handy. Our open-source version comes with a simple way to do it. When you create your user certificate you need to pass the `--add-user` flag, (e.g. `step ssh login --add-user john@doe.com`). This automatically creates another certificate for a user `provisioner` that will run the command `sudo useradd -m <principal>; nc -q0 localhost 22` in the server, create the new user, and proxy the connection to you. This is configurable, and of course, the `provisioner` user must be configured in your host.

In this blog post, you can find some helpful instructions for some of the above, as well as a handy script to bootstrap a server in AWS, but you should be able to adapt to your needs.
https://smallstep.com/blog/diy-single-sign-on-for-ssh/

**Comment by mshamash** [comment, marked as answer]
Hi @maraino and thanks for the detailed reply. Since editing /etc/ssh/sshd_config is a must for SSH certificates to function properly, it looks like I won't be able to use it on the majority of servers where I'm currently a user (one of many) and not an admin with authority to edit global SSH configs.

Hopefully certificate authentication for SSH becomes more widespread and adopted by our compute cluster provider!

**Reply by maraino** [reply]
It's a shame, it makes the SSH key management way easier. For example, you won't need to distribute keys for new employees, with the OAuth/OIDC provisioner, as long as you add them to your IdP, and add them to the right groups he will have access using SSH certificates.

**Reply by mmalone** [reply]
If all you're looking for is the ability to authenticate using single sign-on we may be able to make this work.

When you SSH to a server, _mutual_ authentication occurs: the server authenticates you and you authenticate the server. Our instructions describe using certificates for both, but user and host authentication can be configured to use different mechanisms.

As @maraino mentioned, the `HostCertificate` and `HostKey` directives in `/etc/ssh/sshd_config` are setting up certificate authentication for the host. The `TrustedUserCAKeys` directive, on the other hand, enables certificate authentication for users: it tells the host "allow users to authenticate using a certificate issued by this CA" (the CA is identified by its public key). So, if you only set `TrustedUserCAKeys`, you'll only enable cert authentication for users, which means hosts don't need a cert and you don't need to setup the daemons for cert renewal, etc.

Of course, using `TrustedUserCAKeys` still requires editing `/etc/ssh/sshd_config`. This directive enables certificate authentication for users system-wide. Fortunately, you can also enable certificate authentication for an individual user by using a `@cert-authority` directive in `~/.ssh/authorized_keys` for that user, which does not typically require privileged access on the machine. For example:

```
@cert-authority ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbm...
```
The `ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbm...` portion of the line should be the same thing that'd ordinarily be written to `/etc/ssh/ca.pub`. So you can copy-paste it from another server that you've already setup.

The only catch is that, on some systems, sysadmins have locked down `~/.ssh/authorized_keys` or they relocate the authorized keys settings to a central directory that only root can edit. If that's the case, you might be out of luck unless you can convince the admin to set a `@cert-authority` directive for you.

Be careful futzing around in `~/.ssh/authorized_keys`. If you delete your key(s) you won't be able to connect back to the server. I recommend keeping a connection open to the server, and testing in a new window, so you can revert if something goes sideways. If you don't have a `~/.ssh/authorized_keys` file and you end up having to make one keep in mind that OpenSSH is very finicky about permissions on files under `~/.ssh`. The files should be owned by your user and the correct modes are:

```
chmod 700 ~/.ssh
chmod 644 ~/.ssh/authorized_keys
```

I wrote this from memory and I haven't tried this in a while, so I might be remembering some details wrong. If you run into any trouble lemme know and I'll see if I can help debug. Good luck!

**Reply by mshamash** [reply]
Oh thanks @mmalone that sounds like exactly what I’d need for most of the servers I connect to! I’ll give it a shot this week and report back with my findings. I’m guessing the `AAAAE2VjZHNhLXNoYTItbm...portion` is just my intermediate CA’s public key which I can get from my CA itself? 

Very excited to give this a go.

**Reply by mmalone** [reply]
@mshamash awesome!

The `AAAAE2VjZHNh...` portion is your SSH User CA public key. OpenSSH uses a different certificate format (it doesn't use x.509 certs) and OpenSSH certs do not support intermediate certificates. When you run `step ca init --ssh` we generate an OpenSSH User & Host CA public key for you, in addition to your x.509 root & intermediate. This file lives in `~/.step/certs/ssh_user_ca_key.pub` normally. You can also retrieve the public key from your CA by running `step ssh config --roots`.

That said, I see you mentioned you're running a "Tiny CA". I'm not sure whether the Tiny CA instructions actually generate OpenSSH CA keys. @maraino or @tashian may know.

**Reply by mshamash** [reply]
Gotcha, I don't think it did, since I didn't include the `--ssh` (hindsight is 20/20!). But I came across #400 and will follow @maraino's suggestions of adding a couple lines to ca.json as well as making SSH CA keys with my root CA password and:

```
step crypto keypair ssh_host_ca.pub ssh_host_ca_key
step crypto keypair ssh_user_ca.pub ssh_user_ca_key
```

**Reply by maraino** [reply]
That will give you the key in PEM format, to convert it to SSH just run:
```
step crypto key format --ssh ssh_user_ca.pub
```

**Reply by mshamash** [reply]
I'm wondering if it's "safe" storing the secrets (ssh_host_ca_key & ssh_user_ca_key) on device? For my main certificate's private keys I stored them on my Yubikey's PIV interface (slots 9a & 9c) but don't know if that's an option here considering those slots are now in use (unless I get another Yubikey!)...forgive my naivety, I am still dipping my toes in the water with all this!

Also, the password for the keypairs I generate with `step crypto keypair` command must be the same as my root CA password which I already generated? Or can be another password specifically for SSH keys which I will note down separately? I'm not using a password.txt file

**Reply by maraino** [reply]
You can try to use another slot in the yubikey, create the private key there using ykman or yubico-piv-tool and extract the public key that then you will convert to the OpenSSL format.
https://developers.yubico.com/PIV/Introduction/Certificate_slots.html

You can configure the SSH signing using yubikey like this (of course setting the right slots):
```json
{
        "...": "...",
        "ssh": {
                "hostKey": "yubikey:slot-id=9x",
                "userKey": "yubikey:slot-id=9y"
        },
        "kms": {
                "type": "yubikey",
                "pin": "123456"
        },
        "...": "..."
}
```

**Reply by mshamash** [reply]
Good news, it worked!! I generated the private SSH Host and SSH User keys on the yubikey itself (slots 9d and 9e) with the yubico-piv-tool. Then I extracted the public keys (`yubico-piv-tool -s9e -aread-cer`) and converted to ssh format with `step crypto key format --ssh ssh_user_ca_key.pub`. Finally, on my host (server I'm connecting to), I edited the `~/.ssh/authorized_keys` file to include the following line (my SSH user public key generated on the yubikey):

`cert-authority,principals="michael@mydomain.tld" ecdsa-sha2-nistp256 AAAAE2VjZHNhLX...`

Then I logged in with SSO on my local machine (`step ssh login michael@mydomain.tld --provisioner "Google"`) which I assume I'll need to do every 18 hours (unless I edit the expiry time)...and voila! Secure, passwordless authentication with my own CA. Thanks so much for all your help!

At this point, I'm wondering what the SSH Host key is for then, since it seems like the user key is what's being used throughout authentication? My understanding is the SSH Host key is what signs my 18-hour certificate on my local machine once I pass SSO?
Next I'll be looking into ways to automate the `step ssh login michael@mydomain ....` step so I don't have to run it every day, but can have it run instead upon my first `ssh user@hostname` command of the day.

**Reply by tashian** [reply]
@maraino I had no idea the SSH CA could also use YubiKey PIV! That's so cool.

**Reply by maraino** [reply]
> At this point, I'm wondering what the SSH Host key is for then, since it seems like the user key is what's being used throughout authentication? My understanding is the SSH Host key is what signs my 18-hour certificate on my local machine once I pass SSO?

The host key is used to sign host certificates, you don't need it if you are not going to use certificates in hosts. Using certificates on hosts allows you to have just one `@cert-authority` entry in your known_host and trust automatically hosts with certificates signed by the host key.

> Next I'll be looking into ways to automate the step ssh login michael@mydomain .... step so I don't have to run it every day, but can have it run instead upon my first ssh user@hostname command of the day.

There is a way to do this, if you look at our default config template: https://github.com/smallstep/certificates/blob/00c6f0861267a679a2a1c6f670839f26ff91c718/templates/values.go#L81-L91
The `step ssh proxycommand` can detect if you need a new certificate, looking at the keys in the ssh agent, and do the OIDC flow for you. With more details:

```
# ssh client rule to match the hosts, you can also use things like `Host *`, `Host 192.168.*`, see `man ssh_client`:
Match exec "step ssh check-host %h"
        # Default username to use, by default ssh use the system username
	User mariano
	# Path the the known_host where @cert-authority leaves (only needed if you use host certificates)
	UserKnownHostsFile "/path/to/ssh/known_hosts"
	# Get the bastion configuration from the ca, and uses `Google` provisioner to start an automatic `step ssh login`, and then proxies the connection to your client.
	ProxyCommand step ssh proxycommand --provisioner "Google" %r %h %p
```

**Reply by mshamash** [reply]
I see, I've changed the `Match exec "step ssh check-host %h"` line to `Host *` so it'll trigger SSO/cert creation for any host I connect to, and am just realizing that the `step ssh proxycommand` requires that the remote username be equal to one of the Principals on my certificate (so `michael` or `michael@mydomain.tld`). Is there a way around this with `--set=<key=value>` perhaps?

**Reply by maraino** [reply]
The best work around is using certificate templates.

**Reply by maraino** [reply]
Although there's another one using ssh's `AuthorizedPrincipalsCommand` or `AuthorizedPrincipalsFile` and use the email address in your certificate

**Reply by mshamash** [reply]
I've added the following block to my ca.json file so that I can edit the default SSH cert template (from [here](https://github.com/smallstep/crypto/blob/162770cad29063385cb768b0191814e4c6a94e45/sshutil/templates.go#L144))

```
"options": {
    "ssh": {
        "templateFile": "/etc/step-ca/templates/ssh/cert.tpl"
    }
}
```

As of now my `cert.tpl` file looks like this (replicating the default file):
```
{
        "type": "{{ .Type }}",
        "keyId": "{{ .KeyID }}",
        "principals": ["michael@mydomain.tld"],
        "extensions": {{ toJson .Extensions }},
        "criticalOptions": {{ toJson .CriticalOptions }}
}
```

~~How can I set the `principals` portion to include the .Principals obtained from OIDC (so `michael` and `michael@mydomain.tld`) **and** some custom defined ones (ie. other usernames on servers I connect to, like `ubuntu`, `mike`, etc.)? I'm not familiar with this format of the cert.tpl file. Thanks!~~

~~EDIT: I've been playing around some more, and even if I replace the `"principals": {{ toJson .Principals }},` line with `"principals": ["ubuntu"]`, I still get a step-ca server error of `authority.SignSSH: ssh certificate principals does not match - got [ubuntu], want [michael michael@mydomain.tld]`. This is with `Host * ` set in my step client SSH config, and me connecting to a host with `ssh ubuntu@serverIP`~~

~~EDIT2 (Jan 10): I added myself as an admin in the ca.json file for my Google OIDC provider by adding the `"admins": ["michael@mydomain.tld"]` line. Now I'm able to issue SSH certs for any host/Principal pair I connect to. The only thing I don't like is how I'm forcing the `michael@mydomain.tld` for the principals on the cert (see my cert.tpl above). Without it, then the only principal on the SSH cert if the username of the host I connect to, which means my SSH cert will ONLY work on that host. Some hosts have different usernames, so this isn't feasible, that's why I'm forcing a principal of my email address.~~
~~**Is there any way to set the cert.tpl file Principal to be ONLY my email address as provided by the OIDC provider, instead of hard-coding it in?**~~

**Final edit!**: Turns out changing the principals line in the template to `"principals": ["{{ .Token.email }}"]` did the trick! Thanks again for all your help over the past week!
Cheers. Michael
